Add run all test support with define REDIS_TEST

sundb opened the pull request:
1. Add `redis-server test all` support to run all tests.
2. Move dict benchmark to REDIS_TEST.

oranagra commented on src/Makefile:
```diff
@@ -277,9 +277,9 @@ REDIS_SERVER_NAME=redis-server$(PROG_SUFFIX)
 REDIS_SERVER_OBJ=adlist.o quicklist.o ae.o anet.o dict.o server.o sds.o zmalloc.o lzf_c.o lzf_d.o pqsort.o zipmap.o sha1.o ziplist.o release.o networking.o util.o object.o db.o replication.o rdb.o t_string.o t_list.o t_set.o t_zset.o t_hash.o config.o aof.o pubsub.o multi.o debug.o sort.o intset.o syncio.o cluster.o crc16.o endianconv.o slowlog.o scripting.o bio.o rio.o rand.o memtest.o crcspeed. ...
 REDIS_CLI_NAME=redis-cli$(PROG_SUFFIX)
-REDIS_CLI_OBJ=anet.o adlist.o dict.o redis-cli.o zmalloc.o release.o ae.o crcspeed.o crc64.o siphash.o crc16.o monotonic.o cli_common.o mt19937-64.o
+REDIS_CLI_OBJ=anet.o adlist.o dict.o redis-cli.o zmalloc.o release.o ae.o crcspeed.o crc64.o siphash.o crc16.o monotonic.o cli_common.o mt19937-64.o sds.o
```
@sundb why did you add `sds.o` to redis-cli and redis-benchamrk?
in this past, this was actually very dangerous since these applications contains the sds implementation that's inside hiredis, and the symbols can clash.
nowadays, the sds methods inside hiredis have `hi_` prefix so the symbols don't clash, but redis-cli still attempts to use sds methods on strings allocated by hiredis (using macros provided by sdscompat.h), had we also added an include to redis's sds.h that could have cause a mess.

bottom line, these programs should not currently use redis's sds implementation, and whatever you were trying to solve, we need to solve differently.

sundb replied:
I used `#include <sds.h>` in dict  tests to solve the problem.

oranagra commented on src/server.c:
```diff
@@ -5913,36 +5913,54 @@ int iAmMaster(void) {
+        if (!strcasecmp(argv[2], "all")) {
+            int numtests = sizeof(redisTests)/sizeof(struct redisTest);
+            for (int j = 0; j < numtests; j++) {
+                redisTests[j].proc(argc,argv);
```
here we need to either return 1 if the proc returned non-zero.
or better yet, keep a list of failed tests and print it at the end (returning 1 if any of them returned non-zero)

sundb replied:
Done it.

oranagra reviewed (commented):
please add this to the daily CI (`.github/workflows/daily.yml`).
can be a last step of both `test-ubuntu-jemalloc` and `test-ubuntu-32bit`.
i would also like to copy the use of `-Werror` to both of these, to make sure there are no complication warnings when `-DREDIS_TEST` is added.

oranagra commented:
p.s. in order to test this as part of this PR, we can temporarily copy these portions of `test-ubuntu-jemalloc` and `test-ubuntu-32bit` from daily.yml into their counterparts in ci.yml, and remove before merging.

sundb commented:
I have added tests to ci.yml.

zuiderkwast commented on src/dict.c:
```diff
@@ -1216,14 +1213,14 @@ dictType BenchmarkDictType = {
 } while(0)
 
 /* dict-benchmark [count] */
-int main(int argc, char **argv) {
+int dictTest(int argc, char **argv) {
```
Please update the comment. The executable is not `dict-benchmark` anymore and `count` is not in argv[1] anymore, apparently.

sundb replied:
Got it.

oranagra commented on .github/workflows/ci.yml:
```diff
@@ -58,3 +58,27 @@ jobs:
+    steps:
+    - uses: actions/checkout@v2
+    - name: make
+      run: make REDIS_CFLAGS='-DREDIS_TEST'
```
please also add `-Werror` to both, let's see that they pass cleanly

oranagra commented on src/dict.c:
```diff
@@ -1173,7 +1173,8 @@ void dictGetStats(char *buf, size_t bufsize, dict *d) {
 #ifdef REDIS_TEST
 
-#include "sds.h"
+#include <sdscompat.h>
+#include <sds.h>
```
this can cause some issues, bringing hiredis's sds into redis-server.
IIUC, the problem is that when building redis-cli, we take the dict.o file that was compiled with `-DREDIS_TEST`.
i think this is an indication that we need to do one of the following:
1. avoid using dict.c in redis-cli (obviously not an option).
2. compile 2 different copies of dict.o, the one used by redis-cli should be one that doesn't have `-DREDIS_TEST` (a bit complicated)
3. avoid using sds in the dict.c test. (this is probably not too hard)
4. decide that the dict-benchmark is not actually a test to be part of the REDIS_TEST campaign (sounds logical, this is actually a benchmark, not a test)

i think the easiest thing is to go with 4, and just revert the changes to the Makefile and dict.c

sundb replied:
@oranagra I try to go with 3.

oranagra commented:
@sundb thanks. all looks good.
next step should be to remove the temporary code from `ci.yml` and add it to `daily.yml`.
but then i'm worried that since we don't compile this code in the CI, some PR will break it, or add warnings, and then the daily tests will fail.

@yossigo do you think we should run or at least build this code in ci.yml too?

the tests themselves seem to take about 2 minutes, but they do produce massive amount of prints that cause the action view of the test to be slow, so we may want to tone it down.

yossigo commented:
@oranagra these are code paths that don't change frequently, so a daily might be enough.
We do need to silence the prints though, and maybe also run once with valgrind.

sundb commented:
@oranagra I have moved the ci to daily.yml.
@yossigo I have run valgrind and fixed some problem, there are still some warnings that need further work.

sundb commented:
@oranagra In the quicklistRotate method, when the value is char*, there will be a memory overlap problem.
Although this problem will not be wrong, `memcpy(p,s,slen)`, the memory addresses of p and s are overlapped, but the copy is never overlapped.
But valgrind will have a warning, I am not sure if this problem needs to be handled.

quicklistRotate -> quicklistPushHead -> ziplistPush -> __ziplistInsert-> memcpy

oranagra commented:
@sundb i do think it should be addressed (low priority since this is dead code).
i suppose that if Valgrind produces a warning then the memory regions indeed overlap, and i suppose the right thing to do is detect it in quicklistRotate and fallback to a different approach when they overlap (copy to a temp buffer and use it from there).

oranagra commented on .github/workflows/daily.yml:
```diff
@@ -261,3 +261,27 @@ jobs:
     - name: cluster tests
       run: ./runtest-cluster
+
+  redis-test-ubuntu-jemalloc:
+    runs-on: ubuntu-latest
+    if: github.repository == 'redis/redis'
+    timeout-minutes: 14400
+    steps:
+    - uses: actions/checkout@v2
+    - name: make
+      run: make REDIS_CFLAGS='-Werror -DREDIS_TEST'
+    - name: test
+      run: ./src/redis-server test all
+
+  redis-test-ubuntu-32bit:
+    runs-on: ubuntu-latest
+    if: github.repository == 'redis/redis'
+    timeout-minutes: 14400
+    steps:
+    - uses: actions/checkout@v2
+    - name: make
+      run: |
+        sudo apt-get update && sudo apt-get install libc6-dev-i386
+        make 32bit REDIS_CFLAGS='-Werror -DREDIS_TEST'
+    - name: test
+      run: ./src/redis-server test all
```
i think these can be merged into the existing jobs.
i.e. find 3 of the jobs that build and test redis (normal, 32bit and valgrind), add the CFLAGS, and an extra step at the end.

sundb replied:
I will make the change tomorrow.

sundb commented:
@oranagra Now there is still a problem of silence the prints, I try to use some ways to solve this problem, but always feel unsatisfied.
Move the print interface to testhelp.h
1. add REDIS_TEST_QUIET macro
2. add --quite flag
But both of these ways will change the code a lot, do you have any suggestions?

One more question, run with valgrind is very very slow, especially the ziplist test, I have not finished executing it after half a day.

sundb commented:
@oranagra I'm trying to use the original REDIS_TEST_VERBOSE macro to disable printing, is this ok? This is only a partial change, I will make further changes if possible.

oranagra commented on src/quicklist.c:
```diff
@@ -1296,17 +1296,24 @@ void quicklistRotate(quicklist *quicklist) {
-    unsigned char *value;
+    unsigned char *value, *tmp;
     long long longval;
     unsigned int sz;
     char longstr[32] = {0};
-    ziplistGet(p, &value, &sz, &longval);
+    ziplistGet(p, &tmp, &sz, &longval);
 
     /* If value found is NULL, then ziplistGet populated longval instead */
-    if (!value) {
+    if (!tmp) {
         /* Write the longval as a string so we can re-add it */
         sz = ll2string(longstr, sizeof(longstr), longval);
         value = (unsigned char *)longstr;
+    } else if (quicklist->len == 1) {
+        /* Copy buffer since there could be an memory overlap when move
+         * entity from tail to head in ziplist. */
+        value = zmalloc(sz);
+        memcpy(value, tmp, sz);
+    } else {
+        value = tmp;
```
i think we better move this fix to a PR of it's own.
despite being dead code, i rather not hide it inside the mass of change related to the tests.

sundb replied:
OK.

oranagra commented:
@sundb obviously we don't wanna disable all printing, but we do need to drastically tone it down.
one way to do it, and also make the valgrind run shorter, is to drastically reduce some loops.
maybe something like the `--accurate` argument the test suite has, so in the valgrind mode and maybe in every CI, there's no need to run though 1000 random iterations of some tests, and instead run only 10.

sundb commented:
@oranagra I added the accurate parameter to redisTestProc to reduce the number of iterations, which didn't feel good enough, but I didn't think of any other way to pass this parameter.
Run valgrind now takes about 4 minutes.

oranagra commented on src/quicklist.c:
```diff
@@ -1715,7 +1716,8 @@ int quicklistTest(int argc, char *argv[]) {
     long long runtime[option_count];
 
-    for (int _i = 0; _i < (int)option_count; _i++) {
+    iteration = accurate ? (int)option_count : 1;
+    for (int _i = 0; _i < iteration; _i++) {
         printf("Testing Option %d\n", options[_i]);
```
setting `iteration` to 0 means we don't test compression, right?
would limiting it to [0,1.2] be too slow?
```suggestion
    iteration = accurate ? (int)option_count : 3;
    for (int _i = 0; _i < iteration; _i++) {
        printf("Testing Compression option %d\n", options[_i]);
```

sundb replied:
setting iteration to 0 means all tests are not run, similar, 2 means test compress from 0 to 2, 
I'll test how much the speed has changed.

oranagra replied:
ohh, sorry, for some reason i thought we run only one iteration (the first one with no compression).
(not enough coffee in my blood).
i don't think we want to completely skip this test, we i think we need to run at least with and without compression), and find other ways to tone down the prints and time.

sundb replied:
I try it.

oranagra commented on .github/workflows/daily.yml:
```diff
@@ -28,6 +28,8 @@ jobs:
     - name: cluster tests
       run: ./runtest-cluster
+    - name: unittest
+      run: ./src/redis-server test all --accurate
```
i was hoping to use this `accurate` feature to reduce the number of prints.
so either we remove the `--accurate` from here, or find a few of the tests that print the most, and tone them down.

sundb replied:
Got it.

oranagra reviewed (commented):
maybe we should copy these 3 jobs back to `ci.yml` again so the result is visible in the PR page (and drop them again before merging)

sundb commented:
@oranagra I made a few changes.
1. Remove TEST_DESC from the loop reduces the amount of printing, with the disadvantage of not print the fill.
2. Remove OK macro, feels like it is pointless and will cause a lot of print.
3. Add a fixed fill array to speed up testing.

The total test time is 4 minutes.

oranagra commented on .github/workflows/ci.yml:
```diff
@@ -113,12 +91,5 @@ jobs:
-    - name: test
-      run: |
-        sudo apt-get update
-        sudo apt-get install tcl8.6 valgrind -y
```
you need to add this back

oranagra commented on src/quicklist.c:
```diff
@@ -1739,54 +1741,52 @@ int quicklistTest(int argc, char *argv[], int accurate) {
-        for (int f = optimize_start; f < 32; f++) {
-                quicklist *ql = quicklistNew(f, options[_i]);
+        for (int f = 0; f < fill_count; f++) {
+                quicklist *ql = quicklistNew(fills[f], options[_i]);
```
let me check if i understand, these loops were going from -5 to 32 (total of 37 iterations), and now they run 18 iterations, using values from -5 to 1024?
is `66` a typo?
not sure we need that many iterations on the high end.
i guess we can tone it down further more to some 10 or 12 iterations?
i guess there is some value of some values that are non-power of two?

sundb replied:
1. Yes, now only up to 18 iterations.
2. `66` is to cover the original test
```
if (fills[f] == 1)
    ql_verify(ql, 66, 66, 1, 1);
else if (fills[f] == 32)
    ql_verify(ql, 3, 66, 32, 2);
else if (fills[f] == 66)
    ql_verify(ql, 1, 66, 66, 66);
```
3. Since the old fill maximum is 1024, I'll remove and add some special fills.

oranagra replied:
ohh, i missed the fact that some iterations were running till 72 and 1024 (thought they all run from -5 to 32)

oranagra commented on src/quicklist.c:
```diff
@@ -1739,54 +1741,52 @@ int quicklistTest(int argc, char *argv[], int accurate) {
                 if (ql->count != 5)
                     ERROR;
-                if (f == 32)
+                if (fills[f] == 32)
```
i guess this was meant to be run in the last iteration (`f==fill_count`), not when the value is 32.
am i wrong?

sundb replied:
I found out that the original test will never go in, because f will never equal 32.
The original test wanted to verify that the length of quicklist doesn't become 2 when fill>5, and doesn't necessarily need to run to the end, but I think it could be changed to `if (fils[f] >= 5)`

oranagra commented:
@sundb thank you. it all looks good.
i see valgrind detects two leaks, and the tests succeeded anyway.
i guess we need to take a similar approach to what the tests are doing, redirect valgrind stderr to a log file, and then search `at 0x` in it (indicating a stack trace), if found, print the log file to the console, and terminate with an error.
please do that first to see that the test ends with an error, and then look into the leak.

sundb commented:
@oranagra Yes, I'm looking into this, and strangely enough, it doesn't happen on my machine.

sundb commented:
@oranagra I made a very silly mistake, because there is `github.repository == 'redis/redis'`, so the test is the merged code, I have been testing on my own branch, and also installed ubuntu 20.04 virtual machine, the fact is that the new code added to the unstable branch has a memory leak.

sundb commented on .github/workflows/ci.yml:
```diff
@@ -58,3 +58,44 @@ jobs:
+    - name: unittest
+      run: |
+        valgrind --track-origins=yes --suppressions=./src/valgrind.sup --show-reachable=no --show-possibly-lost=no --leak-check=full ./src/redis-server test all 2>> err.txt
+        if grep -q 0x err.txt; then cat err.txt; exit 1; fi
```
Is it OK? But does cat result in print log twice.

oranagra replied:
why would cat print it twice?

p.s. maybe we wanna use `--log-file=` option for valgrind, instead of `2>` . not sure if the tests print anything meaningful to stderror

oranagra commented:
> because there is github.repository == 'redis/redis', so the test is the merged code,

are you sure? AFAIK it only means that the test is skipped, it doesn't affect the repo the code is pulled from, and doesn't perform any implicit merges. am i wrong? can you provide some proof or reference?

sundb commented:
```
==5116== 3,744,000 bytes in 58,500 blocks are definitely lost in loss record 3 of 3
==5116==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==5116==    by 0x161BE4: ztrymalloc_usable (zmalloc.c:101)
==5116==    by 0x161C4B: zmalloc (zmalloc.c:119)
==5116==    by 0x1434CF: _quicklistSaver (quicklist.c:1397)
==5116==    by 0x143434: quicklistPopCustom (quicklist.c:1378)
==5116==    by 0x14355D: quicklistPop (quicklist.c:1414)
==5116==    by 0x149716: quicklistTest (quicklist.c:2672) <- error line num.
==5116==    by 0x15C6AE: main (server.c:6080)
```

When I merged the latest unstable, the line numbers were correct.

oranagra commented:
you mean that the error line number in https://github.com/redis/redis/runs/2065494261?check_suite_focus=true refers to a line your branch didn't have at that time.

@yossigo are you aware of such a thing?
that the following line in our action means that the test is executed on a merge result, rather than just a condition?
```
    if: github.repository == 'redis/redis'
```

sundb commented:
@oranagra Yes, I found the memory leak line after merge.

sundb commented:
![1](https://user-images.githubusercontent.com/965798/110478183-0d2db280-811f-11eb-99fd-fa2f27a0da5e.png)
[https://github.com/redis/redis/runs/2065493698?check_suite_focus=true](https://github.com/redis/redis/runs/2065493698?check_suite_focus=true)
```
HEAD is now at 125ffd6 Merge de4a38c9359e4ebce61dffd2c2b5d80e586f88eb into af2175326c87231d16c5633dd8febbc86a258540
```
It looks like all the tasks will be merged unstable and then executed.
